### PR TITLE
[Merged by Bors] - feat: tensor product of two flat modules is flat

### PR DESCRIPTION
**Pull request by alreadydone**
Add the missing `Flat R (M ⊗[R] N)` instance.

Add two lemmas relating flatness to injectivity of `mapIncl`.

Clean up the file Flat/Stability, renaming `Flat.comp` to `Flat.trans`.

Also rename `Algebra.Flat.comp` and `FaithfullyFlat.comp`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/RingTheory/Flat/Basic.lean [hunk 1]
```diff
@@ -393,6 +398,22 @@ theorem iff_rTensor_exact : Flat R M ↔
...
+  exact (lTensor_preserves_injective_linearMap _ q.injective_subtype).comp
+    (rTensor_preserves_injective_linearMap _ p.injective_subtype)
+
+/-- If p and q are submodules of M and N respectively, and N and p are flat,
+then `p ⊗ q → M ⊗ N` is injective. -/
+theorem tensorProduct_mapIncl_injective' (p : Submodule R M) (q : Submodule R N)
+    [Flat R p] [Flat R N] : Function.Injective (mapIncl p q) := by
+  rw [mapIncl, ← rTensor_comp_lTensor]
+  exact (rTensor_preserves_injective_linearMap _ p.injective_subtype).comp
+    (lTensor_preserves_injective_linearMap _ q.injective_subtype)
```
Comment (alreadydone) [inline Mathlib/RingTheory/Flat/Basic.lean:415]: @acmepjz maybe you already have these somewhere for linear disjointness?

Comment (acmepjz) [reply Mathlib/RingTheory/Flat/Basic.lean:415]: It's very similar to [Submodule.LinearDisjoint.of_le_of_flat_left](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/LinearDisjoint.html#Submodule.LinearDisjoint.of_le_of_flat_left) and [Submodule.LinearDisjoint.of_le_of_flat_right](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/LinearDisjoint.html#Submodule.LinearDisjoint.of_le_of_flat_right) but not completely the same. By the way, in my PR about linear disjoint, the only place mentioning `mapIncl` is [Submodule.mulMap_eq_mul'_comp_mapIncl](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/TensorProduct/Submodule.html#Submodule.mulMap_eq_mul'_comp_mapIncl), I think.

Comment (acmepjz) [reply Mathlib/RingTheory/Flat/Basic.lean:415]: Do you think it's a good idea to name them as `of_flat_[left|right]` instead of `'`?

Comment (alreadydone) [reply Mathlib/RingTheory/Flat/Basic.lean]: Yes! I opt for `of_[left|right]`.

File: Mathlib/RingTheory/Flat/Basic.lean [hunk 2]
```diff
@@ -260,6 +260,11 @@ theorem rTensor_preserves_injective_linearMap {N' : Type*} [AddCommGroup N'] [Mo
 @[deprecated (since := "2024-03-29")]
 alias preserves_injective_linearMap := rTensor_preserves_injective_linearMap
 
+instance [Flat R M] [Flat R N] : Flat R (M ⊗[R] N) :=
+  (iff_rTensor_injective' _ _).mpr fun I ↦ by
+    simpa [rTensor_tensor] using rTensor_preserves_injective_linearMap _
+      (rTensor_preserves_injective_linearMap _ I.injective_subtype)
```
Comment (erdOne) [inline Mathlib/RingTheory/Flat/Basic.lean:266]: Is it easy to get the heterobase version (i.e. `[Flat S M] [Flat R N] : Flat S (M ⊗[R] N)`)

**Review by erdOne — APPROVED** [review]
LGTM otherwise

File: Mathlib/RingTheory/Flat/Basic.lean (hunk 2 above)
Comment (alreadydone) [reply Mathlib/RingTheory/Flat/Basic.lean]: Done!

**Comment by erdOne** [comment]
Thanks!
maintainer merge

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**